pkg/assembler/backends/ent/backend/hasMetadata.go

```diff
@@ -0,0 +1,262 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backend
+
+import (
+	"context"
+	stdsql "database/sql"
+
+	"entgo.io/ent/dialect/sql"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent/hasmetadata"
+	"github.com/guacsec/guac/pkg/assembler/backends/ent/predicate"
+	"github.com/guacsec/guac/pkg/assembler/backends/helper"
+	"github.com/guacsec/guac/pkg/assembler/graphql/model"
+	"github.com/pkg/errors"
+	"github.com/vektah/gqlparser/v2/gqlerror"
+)
+
+func (b *EntBackend) HasMetadata(ctx context.Context, filter *model.HasMetadataSpec) ([]*model.HasMetadata, error) {
+	records, err := b.client.HasMetadata.Query().
+		Where(hasMetadataPredicate(filter)).
+		Limit(MaxPageSize).
+		WithSource(withSourceNameTreeQuery()).
+		WithArtifact().
+		WithPackageVersion(withPackageVersionTree()).
+		WithAllVersions(withPackageNameTree()).
+		All(ctx)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return collect(records, toModelHasMetadata), nil
+}
+
+func (b *EntBackend) IngestHasMetadata(ctx context.Context, subject model.PackageSourceOrArtifactInput, pkgMatchType *model.MatchFlags, hasMetadata model.HasMetadataInputSpec) (*model.HasMetadata, error) {
+	recordID, err := WithinTX(ctx, b.client, func(ctx context.Context) (*int, error) {
+		return upsertHasMetadata(ctx, ent.TxFromContext(ctx), subject, pkgMatchType, hasMetadata)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &model.HasMetadata{ID: nodeID(*recordID)}, nil
+}
+
+func (b *EntBackend) IngestBulkHasMetadata(ctx context.Context, subjects model.PackageSourceOrArtifactInputs, pkgMatchType *model.MatchFlags, hasMetadataList []*model.HasMetadataInputSpec) ([]string, error) {
+	var results []string
+	for i := range hasMetadataList {
+		var subject model.PackageSourceOrArtifactInput
+		if len(subjects.Packages) > 0 {
+			subject = model.PackageSourceOrArtifactInput{Package: subjects.Packages[i]}
+		} else if len(subjects.Artifacts) > 0 {
+			subject = model.PackageSourceOrArtifactInput{Artifact: subjects.Artifacts[i]}
+		} else {
+			subject = model.PackageSourceOrArtifactInput{Source: subjects.Sources[i]}
+		}
+		hm, err := b.IngestHasMetadata(ctx, subject, pkgMatchType, *hasMetadataList[i])
+		if err != nil {
+			return nil, gqlerror.Errorf("IngestBulkHasMetadata failed with element #%v with err: %v", i, err)
+		}
+		results = append(results, hm.ID)
+	}
+	return results, nil
+}
+
+func hasMetadataPredicate(filter *model.HasMetadataSpec) predicate.HasMetadata {
+	predicates := []predicate.HasMetadata{
+		optionalPredicate(filter.ID, IDEQ),
+		optionalPredicate(filter.Since, hasmetadata.TimestampGTE),
```

Comment:
the other backends are storing time in UTC (or at least should be). Does that need to happen here?

Reply:
thanks for the head up!
done in https://github.com/guacsec/guac/pull/1365/commits/4f5a724e613d7a1fb31de69a31145c55fcc43857#diff-c931a8bfa5cac970a0928d2644523941c98fcfe88b306e77182a4a92e6e2b7f6R112